docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+
+[TIP]
+====
+Your can use programmatic lookup via  `javax.enterprise.inject.Instance` to resolve ambiguities at runtime and even iterate over all beans implementing a given type:
+
+[source,java]
+----
+public class Translator {
+
+    @Inject
+    Instance<Dictionary> dictionaries; <1>
+    
+    String translate(String sentence) { 
+      for (Dictionary dict : dictionaries) { <2>
+         // ...
+      }
+    }
+}
+----
+<1> This injection point will not result in an ambiguous dependency even if there are multiple beans that implement the `Dictionary` type.
+<2> `javax.enterprise.inject.Instance` extends `Iterable`.
+====
+
+== _Q: Can I use setter and constructor injection?_
+
+A: Yes, you can.
+In fact, in CDI the "setter injection" is superseded by more powerful https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#initializer_methods[initializer methods, window="_blank"].
+Intializers may accept multiple parameters and don't have to follow the JavaBean naming conventions.
+
+.Initialized and Constructor Injection Example
+[source,java]
+----
+@ApplicationScoped
+public class Translator {
+
+    private final TranslatorHelper helper
+
+    Translator(TranslatorHelper helper) { <1>
+       this.helper = helper;
+    }
+
+    @Inject <2>
+    void setDeps(Dictionary dic, LocalizationService locService) { <3>
+      / ...
+    }
+}
+----
+<1> This is a constructor injection. In fact, this code would not work in regular CDI implementations where a bean with a normal scope must always declare a no-args constructor. In Quarkus we detect this problem and "add" such a constructor directly in the bytecode.
+<2> An initializer method must be annotated with `@Inject`. 
+<3> An initializer may accept multiple parameters - each one is an injection point.
+
+== _Q: You talked about some qualifiers?_
+
+A: https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#qualifiers[Qualifiers, window="_blank"] are annotations that help the container to distinguish beans that implement the same type.
+As we already said a bean is assignable to an injection point if it has all the required qualifiers.
+If you declare no qualifier at an injection point the `@Default` qualifier is assumed.
+
+.Bean With Custom Qualifier Example
+[source,java]
+----
+@Superior <1>
+@ApplicationScoped
+public class SuperiorTranslator extends Translator {
+
+    String translate(String sentence) { 
+      // ...
+    }
+}
+----
+<1> `@Superior` is a https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#defining_qualifier_types[qualifier annotation, window="_blank"].
+
+This bean would be assignable to `@Inject @Superior Translator` and `@Inject @Superior SuperiorTranslator` but not in `@Inject Translator`.
+The reason is that `@Inject Translator` is automatically transformed to `@Inject @Default Translator` during typesafe resolution.
+And since our `SuperiorTranslator` does not declare `@Default` only the original `Translator` bean is assignable.
```

Comment:
Should we explain how you create a qualifier?

Reply:
+1

Reply:
I think that's covered in the links to spec. You literally just copy the code it links to.